### DISCUSSION TITLE
Upstream memtest86+ audit / work for shim support ?

### DISCUSSION THREAD

**Discussion by debrouxl** [memtest86plus/memtest86plus#370] [General]
Getting some memtest86+ versions supported by shim is clearly going to save time to users (not having to disable Secure Boot when they want / need to use memtest86+), and is likely to further widen inclusion of memtest86+ onto Linux Live / Install CDs.

In https://github.com/rhboot/shim-review/issues/314 , jak suggested that upstream write a self-audit report, in the "memtest86+ won't harm Secure Boot because ..." direction.
Should we want to proceed further, here's my starting point for thoughts about memtest86+'s attack surface:
* it's written in C with a bit of assembly for startup code, which isn't quite the most secure setup in general, but that's what all of the main general-purpose UEFI-capable OS do as well, and memtest86+ has a smaller code base than the monolithic kernels - not sure about the likes of seL4, though;
* just like the other UEFI-capable OS, it accepts arbitrary user input on the keyboard or kernel command line, but that does not make it a general-purpose OS;
* unlike at least some of the general-purpose UEFI-capable OS, it quickly exits UEFI boot services, in order to maximize the amount of available and thereby testable memory - so after this is done, it is no longer even subject to the attack surface of the UEFI boot services;
* like the usual general-purpose UEFI-capable OS, it parses data exposed by the BIOS, such as memory map + DMI / SMBIOS table + ACPI tables, with some safeguards;
* it enumerates the PCI bus, in order to find a narrow subset of devices relevant to its auxiliary operation - USB HCDs and SMBus controllers, or some types of bridges which provide access to SMBus controller functionality, are the main ones which come to my mind. Such functionality can be at least partially disabled by passing suitable parameters onto the command line;
* as far as USB devices are concerned, it only considers hubs and HID keyboards, and does not try to care about every possible descriptor;
* it currently has no concept of a disk controller (SCSI, IDE/PATA, SATA, SAS, NVMe, etc.), no concept of a USB MSD, and relatedly, no concept of a filesystem;
* it has no concept of NICs or network protocols, either;
* access to PCI devices such as SMBus controllers, bridges or USB HCDs largely matches what the Linux or some *BSD kernels do;
* likewise for access to IMC or EDAC;
* on a secondary branch, there's partial support for setting NX on some memory pages - the low-hanging fruit is done but the code could be improved.

**Comment by martinwhitaker** [comment]
This isn't my area of expertise (or interest), but I'm confused by that shim-review issue. Surely Debian already has a signed version of shim, so if Debian wants to sign their build of memtest86+, they can do it themselves.

We would only need our own signed copy of shim if Sam wants to make the ISO images he builds capable of booting with secure boot enabled.

**Comment by x86fr** [comment]
@debrouxl That's a very good start. The shim review process is a bit complicated (especially for a first submission because understanding the chain of trust and the counter-measures they added for security flaws can be difficult), but I think it's an important step for Memtest86+. The ability to run ISO, USB or PXE directly with Secureboot enabled would be great. The process is explained here: https://github.com/rhboot/shim-review. There is many PRs on their repo to understand how to proceed. As soon as we have a signed shim, we will be able to sign our binaries and chainload them. 

The process for Memtest86+ will be different - potentially faster - than the vast majority of other shim review submission because we won't use it with GRUB or a Linux Kernel, but directly.

One of the required point is to explain how we will keep the key private. We need 2 contacts to share the private key and signing process. I would like to be one of the two, maybe @debrouxl can be the other one if Martin isn't interested by this point? Anyway, I think we all start shim's learning curve from the very bottom. :)

**Comment by debrouxl** [comment]
Yup, we all start shim's learning curve from the beginning.
You do need to be one of the two contacts for the signing process... and if Martin isn't interested, then I guess I need to be the other one :)
You and me don't live _that_ far from each other, although I almost never go to your area... so physically meeting each other, should we need to share something, is doable.
